### ISSUE TITLE
Error handling upgrade request TypeError: Cannot read properties of undefined (reading 'bind')

### ISSUE THREAD

**Issue by nthieu1332002** [vercel/next.js#55802]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Home Single Language
Binaries:
  Node: 18.17.1
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 13.5.3-canary.1
  eslint-config-next: 13.5.2
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which example does this report relate to?

nested-components

### Describe the Bug

After changing Nextjs version from "next": "13.4.19" to "next": "^13.5.3-canary.1", or "next": "^13.5.3", this error occurs:

✓ Compiled /(main)/(routes)/servers/[serverId]/[channelId]/page in 917ms (1911 modules)

Error handling upgrade request TypeError: Cannot read properties of undefined (reading 'bind')
    at DevServer.handleRequestImpl (D:\Web\JS Practice\Nextjs\discord-clone\node_modules\next\dist\server\base-server.js:388:50)

And my sidebar disappear.

### Expected Behavior

Display sidebar.

### To Reproduce

![image](https://github.com/vercel/next.js/assets/73569473/d7e3db2a-98f3-42b1-8444-41419108bed5)

**Comment by balazsorban44** [comment]
Please open an issue with a proper reproduction, using the bug report template.

**Comment by gudvinga** [comment]
I found the problem, but I'm not sure if I should open a new issue with a reproduction.
This error occurs from `ws://localhost:3000/_next/webpack-hmr` when you switch between different versions of applications in development.
@balazsorban44 should I open new issue?

**Comment by davidhellsing** [comment]
Same problem here

**Comment by sably** [comment]
i have this issue, oops!!!

**Comment by MichaelFrieze** [comment]
I also have this issue.

**Comment by mastdeebater** [comment]
Same here

**Comment by truong0vanchien** [comment]
The only solution is downgrading next to version 13.4.12. The command is "npm i next@13.4.12".

**Comment by rahulbhooteshwar** [comment]
@balazsorban44 when will this be available in stable release? 

I see it was closed by you & marked as completed.

**Comment by balazsorban44** [comment]
it wasn't investigated as there was no reproduction provided if you have one please open a new issue with one attached as asked here https://github.com/vercel/next.js/issues/55802#issuecomment-1740229448

Note that we don't monitor closed issues.

**Comment by aetrnm** [comment]
Had the same error, what fixed it was:
1. go to node_modules/next/dist/server/base-server.js
2. line 389
3. replace it with `const origSetHeader = _res && typeof _res.setHeader === 'function' ? _res.setHeader.bind(_res) : null;`

**changing the source code is at your own risk**

**Comment by timsun28** [comment]
I've also had this issue before, and reverting to 13.4.12 did the trick for me as well. 
Now with the release of 14, I am not getting this error any more. 

I've had this issue with the Nextra docs theme, so if this was happening for a different type of project for you,  ymmv ...

Hope everyone can resolve this issue!

**Comment by rfviolato** [comment]
I was having this issue while running a proxy side-by-side the next dev server ([using a custom dev server](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server)), this error would pop-up in a WebSockets proxy.

I found out that I had two issues that was causing the WebSockets connection to fail:

- I had a middleware to create Content Security Policy headers, and that was somehow affecting the connection. The solution was to exclucude proxy routes from the middleware like this:
```tsx
// middleware.ts

export function middleware(request: NextRequest) {
   // ...
}

// Adding the config below fixed it
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - favicon.ico (favicon file)
     */
    '/((?!proxy|favicon.ico).*)',
  ],
};
```

But also having a dynamic route setup, something like this `/app/[pageParam]/page.tsx` and connections with the URL `https://localhost:3000/socket/` was most likely being picked up as a page request, so I modified my proxy address from `/socket` to `/proxy/socket`. 

Now I got it working on `v14.0.1`.
